src/functions-and-arrays.js

```diff
@@ -106,9 +187,42 @@ const matrix = [
   [1, 70, 54, 71, 83, 51, 54, 69, 16, 92, 33, 48, 61, 43, 52, 1, 89, 19, 67, 48]
 ];
 
-function greatestProduct() {}
-
-
+function greatestProduct(matrix) {
+  let resultHorizontal = 0;
+  let resultVertical = 0;
+  let highestResult = 0;
+  for (i = 0; i < matrix.length; i++){
+    for (j = 0; j < matrix.length - 3; j++){
+      resultHorizontal = matrix[j][i] * matrix[j][i+1] * matrix[j][i+2] * matrix[j][i+3];
+      if (resultHorizontal > highestResult) {
+        highestResult = resultHorizontal;
+      }
+      resultVertical = matrix[j][i] * matrix[j+1][i] * matrix[j+2][i] * matrix[j+3][i];
+      if (resultVertical > highestResult) {
+        highestResult = resultVertical;
+      }
+  }}
+  return highestResult;
+};
+
+
+function greatestProductOfDiagonals(matrix) {
+  let resultTowardsBottomRight = 0;
+  let resultTowardsBottomLeft = 0;
+  let highestResult = 0;
+  for (i = 0; i < matrix.length; i++){
+    for (j = 0; j < matrix.length - 3; j++){
+      resultTowardsBottomRight = matrix[j][i] * matrix[j+1][i+1] * matrix[j+2][i+2] * matrix[j+3][i+3];
```

Comment:
this should break when you reach i at the end of matrix.length.

Reply:
because i + 3 when i === matrix.length - 1 shouldn't exist in the array. so will be undefined, so would give you NaN.